Angular: Add angular builder to start + build storybook

ThibaudAV opened the pull request:
Issue: https://github.com/storybookjs/storybook/issues/3995
Other : 
https://github.com/storybookjs/storybook/issues/12058
https://github.com/storybookjs/storybook/issues/12232


To bring Storybook closer to angular, add 2 builders, one to start the application and one for the build

Both builders have a required option `browserTarget` which allows storybook to use the configs usually defined for the browser builder (`@angular-devkit/build-angular:browser`) of Angular. This way "reflects" the one already used by `angular-cli` for the builder `serve`, `extract-i18n`.


## What I did

The builder allows to add a "architect" in angular.json to start or build storybook
config ex :
```
        "storybook": {
          "builder": "@storybook/angular:start-storybook",
          "options": {
            "browserTarget": "angular-cli:build", 
            "staticDir": ["src/assets"]
          }
        },
        "build-storybook": {
          "builder": "@storybook/angular:build-storybook",
          "options": {
            "browserTarget": "angular-cli:build",
            "staticDir": ["src/assets"]
          }
        }
```
cmd : `ng run angular-cli:storybook` or `ng run angular-cli:build-storybook`

With this solution it is possible to have several angular projects using different assets and style



## How to test

- Is this testable with Jest or Chromatic screenshots? 
Yes
- Does this need a new example in the kitchen sink apps?
na
- Does this need an update to the documentation?
maybe

ThibaudAV committed:
feat(angular): add angular builder to start storybook

The builder allows to add a "architect" in angular.json to start storybook
config ex :
```
"storybook": {
 "builder": "@storybook/angular:start-storybook",
  "options": {
  "browserTarget": "angular-cli:build",
  "port": 4400
}
```
cmd : `ng run angular-cli:storybook`

With this solution it is possible to have several angular projects using different assets and style

kroeder commented on app/angular/src/builders/start-storybook/index.spec.ts:
```diff
@@ -0,0 +1,54 @@
+
+  beforeEach(async () => {
+    const registry = new schema.CoreSchemaRegistry();
+    registry.addPostTransform(schema.transforms.addUndefinedDefaults);
```
What is this for? (15 and 16) I do lots of stuff with schematics but never have I seen this in tests, I'm curious 🙂

ThibaudAV replied:
no idea. but not nessesary i remove them !
I was inspired by https://angular.io/guide/cli-builder#testing-a-builder ^^

ThibaudAV replied:
this add all default props set in schema.json to test it 

I was inspired by https://angular.io/guide/cli-builder#testing-a-builder ^^

kroeder reviewed (commented):
LGTM so far 👍

ThibaudAV committed:
fixup! feat(angular): add angular builder to start storybook

ThibaudAV committed:
feat(angular): add angular builder to build storybook

The builder allows to add a "architect" in angular.json to build storybook
config ex :
```
"build-storybook": {
 "builder": "@storybook/angular:build-storybook",
  "options": {
    "browserTarget": "angular-cli:build"
  }
}
```
cmd : `ng run angular-cli:build-storybook`

With this solution it is possible to have several angular projects using different assets and style

shilman committed:
Fix deepscan

shilman reviewed (commented):
Looking great

bjornharvold commented:
Great PR - Should be incorporated into nx.dev workflow asap

ThibaudAV commented:
If I have the time I'll go tease them with a PR 😁
